app/src/main/java/fi/helsinki/opintoni/service/converter/PortfolioConverter.java

```diff
@@ -56,12 +62,47 @@ public PortfolioDto toDto(Portfolio portfolio) {
         portfolioDto.visibility = portfolio.visibility;
         portfolioDto.avatarUrl = avatarImageService.getPortfolioAvatarImageUrl(portfolio.getOwnerId());
         portfolioDto.componentVisibilities = componentVisibilityService.findByPortfolioId(portfolio.id);
+
+        fetchPortfolioComponents(portfolio.id, portfolioDto, componentFetchStrategy);
+
         return portfolioDto;
     }
 
+    private void fetchPortfolioComponents(Long portfolioId,
+                                          PortfolioDto portfolioDto,
+                                          ComponentFetchStrategy componentFetchStrategy) {
+        if(componentFetchStrategy == ComponentFetchStrategy.ALL) {
+            fetchAllComponents(portfolioId, portfolioDto);
+        } else if(componentFetchStrategy == ComponentFetchStrategy.PUBLIC) {
+            fetchPublicComponents(portfolioId, portfolioDto);
+        }
+    }
+
+    private void fetchPublicComponents(Long portfolioId, PortfolioDto portfolioDto) {
+        componentVisibilityService.findByPortfolioId(portfolioId).stream()
+            .filter(visibility -> ComponentVisibility.Visibility.PUBLIC.toString().equals(visibility.visibility))
+            .map(visibility -> PortfolioComponent.valueOf(visibility.component))
+            .forEach(component -> fetchComponentData(portfolioId, portfolioDto, component));
+    }
+
+    private void fetchComponentData(Long portfolioId, PortfolioDto portfolioDto, PortfolioComponent component) {
+        switch(component) {
+            case LANGUAGE_PROFICIENCY:
+                portfolioDto.languageProficiencies = languageProficiencyService.findByPortfolioId(portfolioId);
+                break;
+        }
+    }
+
+    private void fetchAllComponents(Long portfolioId, PortfolioDto portfolioDto) {
+        portfolioDto.languageProficiencies = languageProficiencyService.findByPortfolioId(portfolioId);
+    }
```

Comment:
Voisiko tässä käyttää fetchComponentData(portfolioId, portfolioDto, PortfolioComponent.LANGUAGE_PROFICIENCY)?

Näin portfolioDTO:n propertyjen populoimislogiikka tarvisi olla pelkästään fetchComponentData -funktiossa.